File: apps/android/app/src/main/java/inc/anky/android/core/protocol/AnkyValidator.kt
package inc.anky.android.core.protocol

object AnkyValidator {
    fun validate(text: String): AnkyValidation {
        return try {
            if (text.trim().isEmpty()) throw AnkyParseException("EMPTY_ANKY")
            val parsed = AnkyParser.parse(text)
            val durationMs = AnkyDuration.durationMs(parsed)
            val kind = if (AnkyDuration.isComplete(parsed)) AnkyKind.Complete else AnkyKind.Fragment
            AnkyValidation.Valid(kind = kind, parsed = parsed, durationMs = durationMs)
        } catch (error: Throwable) {
            AnkyValidation.Invalid(error = error.message ?: "INVALID_ANKY")
        }
    }
}

File: apps/ios/Anky/Core/Protocol/AnkyValidator.swift
import Foundation

public enum AnkyKind: String, Equatable {
    case fragment
    case complete
}

public struct AnkyValidation: Equatable {
    public let isValid: Bool
    public let kind: AnkyKind?
    public let isComplete: Bool
    public let parsed: ParsedAnky?
    public let durationMs: Int64
    public let error: String?
}

public enum AnkyValidator {
    public static func validate(_ text: String) -> AnkyValidation {
        do {
            if text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                throw AnkyParseError.emptyAnky
            }

            let parsed = try AnkyParser.parse(text)
            let durationMs = AnkyDuration.durationMs(parsed)
            let isComplete = AnkyDuration.isComplete(parsed)

            return AnkyValidation(
                isValid: true,
                kind: isComplete ? .complete : .fragment,
                isComplete: isComplete,
                parsed: parsed,
                durationMs: durationMs,
                error: nil
            )
        } catch {
            return AnkyValidation(
                isValid: false,
                kind: nil,
                isComplete: false,
                parsed: nil,
                durationMs: 0,
                error: String(describing: error)
            )
        }
    }
}
